tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc

```diff
@@ -0,0 +1,437 @@
+    if (!status.ok()) return status;
+
+    status = parser_.Parse(&ptr_, &tensors, &types);  // Parse val
+    if (!status.ok()) return status;
+
+    remainder_ -= (ptr_ - initial_ptr);
+
+    out_tensors->resize(tensors.size());
+    for (int32_t i = 0; i < tensors.size(); i++)
+      (*out_tensors)[permutation_[i]] = std::move(tensors[i]);
+
+    *end_of_sequence = false;
+    return Status::OK();
+  }
+
+  *end_of_sequence = true;
+  return Status::OK();
+}
+
+Status IgniteDatasetIterator::SaveInternal(IteratorStateWriter* writer) {
+  return errors::Unimplemented(
+      "Iterator for IgniteDataset does not support 'SaveInternal'");
+}
+
+Status IgniteDatasetIterator::RestoreInternal(IteratorContext* ctx,
+                                              IteratorStateReader* reader) {
+  return errors::Unimplemented(
+      "Iterator for IgniteDataset does not support 'RestoreInternal')");
+}
+
+Status IgniteDatasetIterator::Handshake() {
+  int32_t msg_len = 8;
+
+  if (username_.empty())
+    msg_len += 1;
+  else
+    msg_len += 5 + username_.length();
+
+  if (password_.empty())
+    msg_len += 1;
+  else
+    msg_len += 5 + password_.length();
+
+  TF_RETURN_IF_ERROR(client_->WriteInt(msg_len));
+  TF_RETURN_IF_ERROR(client_->WriteByte(1));
+  TF_RETURN_IF_ERROR(client_->WriteShort(protocol_major_version));
+  TF_RETURN_IF_ERROR(client_->WriteShort(protocol_minor_version));
+  TF_RETURN_IF_ERROR(client_->WriteShort(protocol_patch_version));
+  TF_RETURN_IF_ERROR(client_->WriteByte(2));
+  if (username_.empty()) {
+    TF_RETURN_IF_ERROR(client_->WriteByte(null_val));
+  } else {
+    TF_RETURN_IF_ERROR(client_->WriteByte(string_val));
+    TF_RETURN_IF_ERROR(client_->WriteInt(username_.length()));
+    TF_RETURN_IF_ERROR(
+        client_->WriteData((uint8_t*)username_.c_str(), username_.length()));
+  }
+
+  if (password_.empty()) {
+    TF_RETURN_IF_ERROR(client_->WriteByte(null_val));
+  } else {
+    TF_RETURN_IF_ERROR(client_->WriteByte(string_val));
+    TF_RETURN_IF_ERROR(client_->WriteInt(password_.length()));
+    TF_RETURN_IF_ERROR(
+        client_->WriteData((uint8_t*)password_.c_str(), password_.length()));
+  }
+
+  int32_t handshake_res_len;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&handshake_res_len));
+  uint8_t handshake_res;
+  TF_RETURN_IF_ERROR(client_->ReadByte(&handshake_res));
+
+  LOG(INFO) << "Handshake length " << handshake_res_len << ", res "
+            << (int16_t)handshake_res;
+
+  if (handshake_res != 1) {
+    int16_t serv_ver_major;
+    TF_RETURN_IF_ERROR(client_->ReadShort(&serv_ver_major));
+    int16_t serv_ver_minor;
+    TF_RETURN_IF_ERROR(client_->ReadShort(&serv_ver_minor));
+    int16_t serv_ver_patch;
+    TF_RETURN_IF_ERROR(client_->ReadShort(&serv_ver_patch));
+    uint8_t header;
+    TF_RETURN_IF_ERROR(client_->ReadByte(&header));
+
+    if (header == string_val) {
+      int32_t length;
+      TF_RETURN_IF_ERROR(client_->ReadInt(&length));
+      uint8_t* err_msg_c = new uint8_t[length];
```

Comment:
This will leak if the following `client_->ReadData()` call fails. Instead of using `new[]`, do something like this:

```c++
string err_msg;
err_msg.resize(length);
TF_RETURN_IF_ERROR(client_->ReadData(err_msg.data(), length));
```

Please fix this pattern throughout the code.

Reply:
Looks like I can't use `err_msg.data()` because it returns `const` object, but I need to fill it in `ReadData`. I added `gtl::MakeCleanup`, so I think leak is avoided.